myia/infer/infnode.py

```diff
@@ -0,0 +1,215 @@
+from ..utils.info import enable_debug
+from ..utils.misc import ModuleNamespace
+from .algo import Require, RequireAll, Unify, infer
+
+X = data.Generic("x")
+
+
+@dataclass(frozen=True)
+class InferenceFunction:
+    fn: types.FunctionType
+
+
+def inference_function(fn):
+    return data.AbstractAtom({"interface": InferenceFunction(fn)})
+
+
+class SpecializedGraph:
+    def __init__(self, base_graph):
+        self.base_graph = base_graph
+        self.graph = None
+
+    def commit(self, sig):
+        graph = self.base_graph.specialize(sig)
+        if self.graph is None:
+            self.graph = graph
+        elif self.graph is not graph:
+            raise TypeError()
+
+
+def resolve(args, unif):
+    ns, name = args
+    assert ns.is_constant(ModuleNamespace)
+    assert name.is_constant(str)
+    resolved = ns.value[name.value]
+    ct = Constant(resolved)
+    yield Replace(ct)
+    res = yield Require(ct)
+    return res
+
+
+def user_switch(args, unif):
+    cond, ift, iff = args
+    _ = yield Require(cond)  # TODO: check bool
+    ift_t = yield Require(ift)
+    iff_t = yield Require(iff)
+    return data.AbstractUnion([ift_t, iff_t], tracks={})
+
+
+def signature(*arg_types, ret):
+    arg_types = [
+        type_to_abstract(argt) if isinstance(argt, type) else argt
+        for argt in arg_types
+    ]
+
+    return_type = type_to_abstract(ret) if isinstance(ret, type) else ret
+
+    def _infer(inputs, unif):
+        inp_types = []
+        for inp in inputs:
+            inp_types.append((yield Require(inp)))
+        for inp_type, expected_type in zip(inp_types, arg_types):
+            autils.unify(expected_type, inp_type, U=unif)
+        return autils.reify(return_type, unif=unif.canon)
+
+    return inference_function(_infer)
+
+
+class Handler:
+    pass
+
+
+inferrers = {
+    operator.mul: signature(X, X, ret=X),
+    operator.add: signature(X, X, ret=X),
+    operator.sub: signature(X, X, ret=X),
+    operator.neg: signature(X, ret=X),
+    operator.le: signature(X, X, ret=bool),
+    operator.truth: signature(X, ret=bool),
+    basics.return_: signature(X, ret=X),
+    basics.resolve: inference_function(resolve),
+    basics.user_switch: inference_function(user_switch),
+    type: signature(
+        X,
+        ret=data.AbstractStructure([X], tracks={"interface": type}),
+    ),
+    basics.make_handle: signature(
+        data.AbstractStructure([X], tracks={"interface": type}),
+        ret=data.AbstractStructure([X], tracks={"interface": Handler}),
+    ),
+    basics.global_universe_getitem: signature(
+        data.AbstractStructure([X], tracks={"interface": Handler}),
+        ret=X,
+    ),
+}
+
+
+class Replace:
+    def __init__(self, new_node):
+        self.new_node = new_node
+
+
+class InferenceEngine:
+    def __init__(self):
+        self.replacements = defaultdict(dict)
+
+    def __call__(self, node, unif):
+        if repl := self.replacements.get((None, None, node), None):
```

Comment:
Once again you don't want this to work on 3.8 :)

Reply:
You mean 3.7? `:=` was added in 3.8, right?